PhysicsTools/PatUtils/interface/SmearedJetProducerT.h

```diff
@@ -21,370 +17,256 @@
+
+    using JetCollection = std::vector<T>;
+
     public:
+        explicit SmearedJetProducerT(const edm::ParameterSet& cfg):
+            m_enabled(cfg.getParameter<bool>("enabled")),
+            m_debug(cfg.getUntrackedParameter<bool>("debug", false)) {
 
-     GenJetMatcherT(const edm::ParameterSet& cfg, edm::ConsumesCollector && iC)
-       : srcGenJetsToken_(iC.consumes<reco::GenJetCollection>(cfg.getParameter<edm::InputTag>("srcGenJets"))),
-         dRmaxGenJetMatch_(0)
-     {
-       TString dRmaxGenJetMatch_formula = cfg.getParameter<std::string>("dRmaxGenJetMatch").data();
-       dRmaxGenJetMatch_formula.ReplaceAll("genJetPt", "x");
-       dRmaxGenJetMatch_ = new TFormula("dRmaxGenJetMatch", dRmaxGenJetMatch_formula.Data());
-     }
-     ~GenJetMatcherT()
-     {
-       delete dRmaxGenJetMatch_;
-     }
-
-     const reco::GenJet* operator()(const T& jet, edm::Event* evt = 0) const
-     {
-       assert(evt);
-
-       edm::Handle<reco::GenJetCollection> genJets;
-       evt->getByToken(srcGenJetsToken_, genJets);
-
-       const reco::GenJet* retVal = 0;
-
-       double dRbestMatch = 1.e+6;
-       for ( reco::GenJetCollection::const_iterator genJet = genJets->begin();
-	     genJet != genJets->end(); ++genJet ) {
-	 double dRmax = dRmaxGenJetMatch_->Eval(genJet->pt());
-	 //std::cout << "genJetPt = " << genJet->pt() << ": dRmax = " << dRmax << std::endl;
-	 double dR = deltaR(jet.p4(), genJet->p4());
-	 if ( dR < dRbestMatch && dR < dRmax ) {
-	   retVal = &(*genJet);
-	   dRbestMatch = dR;
-	 }
-       }
-
-       return retVal;
-     }
+            m_jets_token = consumes<JetCollection>(cfg.getParameter<edm::InputTag>("src"));
 
-    private:
+            if (m_enabled) {
+                m_rho_token = consumes<double>(cfg.getParameter<edm::InputTag>("rho"));
 
-//--- configuration parameter
-     edm::InputTag srcGenJets_;
-     edm::EDGetTokenT<reco::GenJetCollection> srcGenJetsToken_;
+                m_use_txt_files = cfg.exists("resolutionFile") && cfg.exists("scaleFactorFile");
 
-     TFormula* dRmaxGenJetMatch_;
-  };
+                if (m_use_txt_files) {
+                    std::string resolutionFile = cfg.getParameter<edm::FileInPath>("resolutionFile").fullPath();
+                    std::string scaleFactorFile = cfg.getParameter<edm::FileInPath>("scaleFactorFile").fullPath();
 
-  template <typename T>
-  class JetResolutionExtractorT
-  {
-    public:
+                    m_resolution_from_file.reset(new JME::JetResolution(resolutionFile));
+                    m_scale_factor_from_file.reset(new JME::JetResolutionScaleFactor(scaleFactorFile));
+                } else {
+                    m_jets_algo = cfg.getParameter<std::string>("algo");
+                }
 
-     JetResolutionExtractorT(const edm::ParameterSet&) {}
-     ~JetResolutionExtractorT() {}
+                std::uint32_t seed = cfg.getUntrackedParameter<std::uint32_t>("seed", 37428479);
```

Comment:
the value of the seed will change the result of the output produced by the code. So, the parameter should be tracked.
Also, during execution, you may want to tack on the event number and/or the original jet pt so that this code result is reproducible (the seed would need to be predictably updated per event or per jet).